File: declarative_sqlite/lib/src/builders/schema_builder.dart
import 'package:declarative_sqlite/src/schema/table.dart';
import 'package:declarative_sqlite/src/schema/schema.dart';
import 'package:declarative_sqlite/src/builders/table_builder.dart';
import 'package:declarative_sqlite/src/builders/view_builder.dart';
import 'package:declarative_sqlite/src/sync/hlc.dart';

/// A fluent builder for defining a database schema.
///
/// The `SchemaBuilder` is the entry point for defining the structure of your
/// database, including its tables and views.
///
/// Example:
/// ```dart
/// final schemaBuilder = SchemaBuilder();
///
/// schemaBuilder.table('users', (table) {
///   table.guid('id').notNull(Uuid().v4());
///   table.text('name').notNull('Default Name');
///   table.integer('age').notNull(0);
///   table.key(['id']).primary();
/// });
///
/// schemaBuilder.view('user_names', (view) {
///   view.select('name').from('users');
/// });
///
/// final schema = schemaBuilder.build();
/// ```
class SchemaBuilder {
  final _tableBuilders = <TableBuilder>[];
  final _viewBuilders = <ViewBuilder>[];

  /// Defines a table in the schema.
  ///
  /// The [name] is the name of the table.
  /// The [build] callback provides a [TableBuilder] to define the table's
  /// columns and keys.
  ///
  /// The [name] is not allowed to start with `__` (two underscores) as that is
  /// reserved for system tables.
  void table(String name, void Function(TableBuilder) build) {
    final builder = TableBuilder(name);
    build(builder);
    _tableBuilders.add(builder);
  }

  /// Defines a view in the schema.
  ///
  /// The [name] is the name of the view.
  /// The [build] callback provides a [ViewBuilder] to define the view's
  /// SELECT statement.
  void view(String name, void Function(ViewBuilder) build) {
    final builder = ViewBuilder(name);
    build(builder);
    _viewBuilders.add(builder);
  }

  /// Builds the [Schema] object.
  ///
  /// This should be called after all tables and views have been defined.
  Schema build() {
    final userTables = _tableBuilders
        .where((t) => !t.name.startsWith('__'))
        .map((b) => b.build())
        .toList();
    final systemTables = [
      _buildSystemTableSettings(),
      _buildSystemTableFiles(),
      _buildSystemTableDirtyRows(),
    ];

    return Schema(
      tables: [...userTables, ...systemTables],
      views: _viewBuilders.map((b) => b.build()).toList(),
    );
  }

  Table _buildSystemTableSettings() {
    final builder = TableBuilder('__settings');
    builder.text('key').notNull('_');
    builder.text('value');
    builder.key(['key']).primary();
    return builder.build();
  }

  Table _buildSystemTableFiles() {
    final builder = TableBuilder('__files');
    builder.guid('id').notNull('00000000-0000-0000-0000-000000000000');
    builder.guid('owner_id').notNull('00000000-0000-0000-0000-000000000000');
    builder.text('filename').notNull('default');
    builder.integer('remote_version').notNull(0);
    builder.text('mimetype').notNull('application/octet-stream');
    builder.integer('size').notNull(0);
    builder.key(['id']).primary();
    builder.key(['owner_id', 'filename']).index();
    return builder.build();
  }

  Table _buildSystemTableDirtyRows() {
    final builder = TableBuilder('__dirty_rows');
    builder.text('table_name').notNull('default');
    builder.guid('row_id').notNull('00000000-0000-0000-0000-000000000000');
    builder.text('hlc').notNull(Hlc.min.toString());
    builder.key(['table_name', 'row_id']).primary();
    return builder.build();
  }
}

File: packages/core/src/schema/builders/schema-builder.ts
import type { Schema, DbTable } from '../types';
import { TableBuilder } from './table-builder';

/**
 * Builder for database schemas
 */
export class SchemaBuilder {
  private tableBuilders: TableBuilder[] = [];
  
  /**
   * Define a table in the schema
   */
  table(name: string, build: (table: TableBuilder) => void): this {
    const builder = new TableBuilder(name);
    build(builder);
    this.tableBuilders.push(builder);
    return this;
  }
  
  /**
   * Build the final schema
   */
  build(): Schema {
    // Build user tables
    const userTables = this.tableBuilders
      .filter(t => !t.name.startsWith('__'))
      .map(t => t.build());
    
    // Add system tables
    const systemTables = this.buildSystemTables();
    
    // Compute schema version (hash of schema structure)
    const version = this.computeSchemaVersion([...userTables, ...systemTables]);
    
    return {
      tables: [...userTables, ...systemTables],
      views: [], // Views will be added in a future phase
      version,
    };
  }
  
  /**
   * Build system tables
   */
  private buildSystemTables(): DbTable[] {
    const tables: DbTable[] = [];
    
    // __settings table
    const settingsBuilder = new TableBuilder('__settings');
    settingsBuilder.text('key').notNull('_');
    settingsBuilder.text('value');
    settingsBuilder.key('key').primary();
    tables.push(settingsBuilder.build());
    
    // __files table
    const filesBuilder = new TableBuilder('__files');
    filesBuilder.text('id').notNull('');
    filesBuilder.text('fileset').notNull('');
    filesBuilder.text('filename').notNull('default');
    filesBuilder.text('mime_type').notNull('application/octet-stream');
    filesBuilder.integer('size').notNull(0);
    filesBuilder.text('created_at').notNull('');
    filesBuilder.text('modified_at').notNull('');
    filesBuilder.integer('version').notNull(1);
    filesBuilder.text('storage_path').notNull('');
    filesBuilder.key('id').primary();
    filesBuilder.key('fileset', 'filename').index();
    tables.push(filesBuilder.build());
    
    // __dirty_rows table
    const dirtyRowsBuilder = new TableBuilder('__dirty_rows');
    dirtyRowsBuilder.text('table_name').notNull('');
    dirtyRowsBuilder.guid('row_id').notNull('00000000-0000-0000-0000-000000000000');
    dirtyRowsBuilder.text('hlc').notNull('');
    dirtyRowsBuilder.integer('is_full_row').notNull(1);
    dirtyRowsBuilder.key('table_name', 'row_id').primary();
    tables.push(dirtyRowsBuilder.build());
    
    return tables;
  }
  
  /**
   * Compute a version hash for the schema
   */
  private computeSchemaVersion(tables: DbTable[]): string {
    // Simple hash based on schema structure
    const schemaString = JSON.stringify({
      tables: tables.map(t => ({
        name: t.name,
        columns: t.columns.map(c => ({ name: c.name, type: c.type })),
        keys: t.keys,
      })),
    });
    
    // Simple hash function
    let hash = 0;
    for (let i = 0; i < schemaString.length; i++) {
      hash = ((hash << 5) - hash) + schemaString.charCodeAt(i);
      hash = hash & hash; // Convert to 32-bit integer
    }
    
    return Math.abs(hash).toString(36);
  }
}
